src/algebra/lie_algebra.lean

```diff
@@ -205,16 +209,191 @@ def of_associative_algebra (A : Type v) [ring A] [algebra R A] : lie_algebra R A
+    simpa using this,
+  },
+  rw [eq_comm, ←lie_skew ⁅x, y⁆ z, ←lie_skew ⁅x, z⁆ y, ←lie_skew x z, lie_neg, neg_neg,
+      ←sub_eq_zero_iff_eq, sub_neg_eq_add, lie_ring.jacobi],
+end
+
+/--
+The bracket of a Lie algebra as a bilinear map.
+-/
+def bil_lie : L →ₗ[R] L →ₗ[R] L :=
+{ to_fun := Ad,
+  add    := by { unfold Ad, intros, ext, simp [add_lie], },
+  smul   := by { unfold Ad, intros, ext, simp, } }
+
+/--
+The bracket of a Lie algebra as a morphism of Lie algebras.
+-/
+def bil_lie_morphism : L →ₗ⁅R⁆ module.End R L := { bracket := is_morphism_Ad, ..bil_lie }
 
 end lie_algebra
 
+section lie_subalgebra
+
+variables (R : Type u) (L : Type v) [comm_ring R] [add_comm_group L] [lie_algebra R L]
+
+/--
+A Lie subalgebra of a Lie algebra is submodule that is closed under the Lie bracket.
+This is a sufficient condition for the subset itself to form a Lie algebra.
+-/
+structure lie_subalgebra extends submodule R L :=
+(bracket : ∀ {x y}, x ∈ carrier → y ∈ carrier → ⁅x, y⁆ ∈ carrier)
+
+instance lie_subalgebra_coe_submodule : has_coe (lie_subalgebra R L) (submodule R L) :=
+⟨lie_subalgebra.to_submodule⟩
+
+instance lie_subalgebra_lie_algebra [L' : lie_subalgebra R L] : lie_algebra R L' := {
+  bracket  := λ x y, ⟨⁅x.val, y.val⁆, L'.bracket x.property y.property⟩,
+  lie_add  := by { intros, apply set_coe.ext, apply lie_add, },
+  add_lie  := by { intros, apply set_coe.ext, apply add_lie, },
+  lie_self := by { intros, apply set_coe.ext, apply lie_self, },
+  jacobi   := by { intros, apply set_coe.ext, apply lie_ring.jacobi, },
+  lie_smul := by { intros, apply set_coe.ext, apply lie_smul, } }
+
+end lie_subalgebra
+
+section lie_module
+
+variables (R : Type u) (L : Type v) [comm_ring R] [add_comm_group L] [lie_algebra R L]
+variables (M  : Type v) [add_comm_group M] [module R M]
+
+section prio
+set_option default_priority 100 -- see Note [default priority]
+/--
+A Lie module is a module over a commutative ring, together with a linear action of a Lie algebra
+on this module, such that the Lie bracket acts as the commutator of endomorphisms.
+-/
+class lie_module := (action : L →ₗ⁅R⁆ module.End R M)
+end prio
+
+/--
+Every Lie algebra is a module over itself.
+-/
+instance lie_algebra_self_module : lie_module R L L := ⟨lie_algebra.bil_lie_morphism⟩
+
+/--
+A Lie submodule of a Lie module is a submodule that is closed under the Lie bracket.
+This is a sufficient condition for the subset itself to form a Lie module.
+-/
+structure lie_submodule [lie_module R L M] extends submodule R M :=
+(bracket : ∀ {x m}, m ∈ carrier → lie_module.action R L M x m ∈ carrier)
+
+instance lie_submodule_coe_submodule [lie_module R L M]
+  : has_coe (lie_submodule R L M) (submodule R M) := ⟨lie_submodule.to_submodule⟩
+
+instance lie_submodule_has_mem [lie_module R L M]
+  : has_mem M (lie_submodule R L M) := ⟨λ x M', x ∈ (M' : set M)⟩
+
+instance lie_submodule_lie_module [α : lie_module R L M] [M' : lie_submodule R L M]
+  : lie_module R L M' := ⟨{
```

Comment:
```suggestion
  : lie_module R L M' :=
⟨{
```

Reply:
Also, this should then be merged with the next line, so that `add` follows the brace.